### ISSUE TITLE
Resource Loading Madness

### ISSUE THREAD

**Issue by niloy** [niloy/blog#2]
# Resource Loading Madness

In a ideal web world, a page would only download files that it requires. So,
if there are 5 js files and 2 web pages:

```
    Page1       Page2
    A.js        C.js
    B.js        D.js
    C.js        E.js
```

When `Page1` is viewed, the browser would download `A.js`, `B.js` and `C.js`.
When `Page2` is viewed, the browser would download `C.js`, `D.js` and `E.js`.
But making seperate calls for each file would make the page load very very slow.
So we concatante the sets into 2 files.

```
    Set1.js -> concat(A.js, B.js, C.js)
    Set2.js -> concat(C.js, D.js, E.js)
```

When `Page1` is viewed, we load `Set1.js` and likewise for `Page2`. But it
appears that this kind of optimization is in direct conflict with another
type of optimization i.e. load only files that are needed.

When user first visits `Page1`, `Set1.js` is loaded which contains A,B and C.
Then when the user visits `Page2`, `Set2.js` is loaded. Here the code for
`C.js` is being loaded again. In a ideal case, we should load the subset 
`(D.js, E.js)`. But when you have multiple pages, detecting all possible subsets
is too complex and not feasible.

In conclusion, there are two types of optimization possible:
- Concat all required files into a single file and load in browser
- Only load files that are required for that page

And both optimization together is not possible. Choosing the _concat_ way is
the only optimization that works for now as it makes page load fast. In some
distant future, where loading multiple files would not be so taxing
([SPDY](http://en.wikipedia.org/wiki/SPDY)), we can probably implement the second
optimization.
